### PR TITLE
[Performance] Tune up caches invalidation during product save

### PR DESCRIPTION
**Pull request by kalessil**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Some meta and term APIs are clearing the underlying caches, causing cache re-reads and increasing the number of SQLs in unpredictable ways across WooCommerce core and extensions workflows.

This PR fixes several verified cases in product data stores, primarily on the product save path — reducing per-save DB and cache pressure, which delays performance degradation under concurrent updates during BFCM/flash sale events (managed stock products saving is involved via `wc_update_product_stock`/`wc_update_product_stock_status`).

### Verification

| Change reverted | Simple create | Simple update | Variable create | Variable update | Grouped create | Grouped update | Variation create | Variation update |
  |---|---:|---:|---:|---:|---:|---:|---:|---:|                                                                                                                                                                
  | Trunk | 148 | 48 | 414 | 51 | 151 | 49 | 128 | 55 |                                                                                                                                                      
  | Branch | 129 | 27 | 359 | 28 | 129 | 27 | 110 | 37 |                                                                                                                                                       
  | Δ vs trunk | −12.8% | −43.8% | −13.3% | −45.1% | −14.6% | −44.9% | −14.1% | −32.7% |                                                                                                                       
  | w/o `update_or_delete_post_meta` | 148 (+19) | 44 (+17) | 414 (+55) | 46 (+18) | 149 (+20) | 45 (+18) | 128 (+18) | 54 (+17) |                                                                             
  | w/o `update_version_and_type` (product) | 129 (0) | 31 (+4) | 359 (0) | 33 (+5) | 129 (0) | 31 (+4) | 110 (0) | 37 (0) |                                                                                   
  | w/o `update_prices_from_children` (grouped) | 129 (0) | 27 (0) | 359 (0) | 28 (0) | 131 (+2) | 27 (0) | 110 (0) | 37 (0) |                                                                                 
  | w/o `sync_price` (variable) * | 129 (0) | 27 (0) | 359 (0) | 28 (0) | 129 (0) | 27 (0) | 110 (0) | 37 (0) |                                                                                                
  | w/o `update_version_and_type` (variation) | 129 (0) | 27 (0) | 359 (0) | 28 (0) | 129 (0) | 27 (0) | 110 (0) | 38 (+1) | 

  \* `sync_price` runs on `shutdown` via `wc_deferred_product_sync()` — outside the benchmark's SQL counting window, so the delta is not captured here.

### How to test the changes in this Pull Request:

Pre-requisite: apply this patch so the importer logs the stats for further verification:

```patch
Subject: [PATCH] Patch
---
Index: plugins/woocommerce/includes/admin/importers/class-wc-product-csv-importer-controller.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/plugins/woocommerce/includes/admin/importers/class-wc-product-csv-importer-controller.php b/plugins/woocommerce/includes/admin/importers/class-wc-product-csv-importer-controller.php
--- a/plugins/woocommerce/includes/admin/importers/class-wc-product-csv-importer-controller.php	(revision df1dc8f4fe575b39daa2f9880792a367b60b3ce4)
+++ b/plugins/woocommerce/includes/admin/importers/class-wc-product-csv-importer-controller.php	(date 1784916942000)
@@ -325,6 +325,9 @@
 
 		check_ajax_referer( 'wc-product-import', 'security' );
 
+		$queries_before = $wpdb->num_queries;
+		$time_before    = microtime( true );
+
 		try {
 			$file = wc_clean( wp_unslash( $_POST['file'] ?? '' ) ); // PHPCS: input var ok.
 			self::validate_file_path( $file );
@@ -359,6 +362,10 @@
 			$importer         = self::get_importer( $file, $params );
 			$results          = $importer->import();
 			$percent_complete = $importer->get_percent_complete();
+			wc_get_logger()->info(
+				sprintf( 'batch pos=%s pct=%s%% sqls=%d time=%.3fs', $params['start_pos'], $percent_complete, $wpdb->num_queries - $queries_before, microtime( true ) - $time_before ),
+				array( 'source' => 'wc-product-import' )
+			);
 			$error_log        = array_merge( $error_log, $results['failed'], $results['skipped'] );
 
 			update_user_option( get_current_user_id(), 'product_import_error_log', $error_log );
```

- Navigate to admin area: `Products -> All products`
- Click import and complete the import of at least 10 products (you can export test products and re-import those back into the test system)
- Perform one import on this branch and one on trunk (remember the sequence for further verification)
- Navigate to `WooCommerce -> Status -> Logs`
- Explore `wc-product-import` file content
- Verify reduction in the number of SQL queries

### CODE REVIEW

File: plugins/woocommerce/includes/data-stores/class-wc-product-variable-data-store-cpt.php [hunk 1]
```diff
@@ -913,9 +913,12 @@ public function sync_price( &$product ) {
...
 
-		delete_post_meta( $product_id, '_price' );
-		delete_post_meta( $product_id, '_sale_price' );
-		delete_post_meta( $product_id, '_regular_price' );
+		// Performance note: prefilter with metadata_exists, to avoid unnecessary meta cache invalidations and SQLs.
+		$price_metas = array_filter(
+			array( '_price', '_sale_price', '_regular_price' ),
+			static fn( $meta_key ) => metadata_exists( 'post', $product_id, $meta_key )
+		);
+		array_walk( $price_metas, static fn( $meta_key ) => delete_post_meta( $product_id, $meta_key ) );
```
[the remark this answers is not in the document]
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-variable-data-store-cpt.php:931]: Those are datastores' internal metas. The only guarantee in this case is that the datastore will keep those updated and use them where applicable.

File: plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php [hunk 2]
```diff
@@ -256,10 +256,12 @@ protected function get_props_to_update( $object, $meta_key_to_props, $meta_type
 	 * @return bool True if updated/deleted.
 	 */
 	protected function update_or_delete_post_meta( $object, $meta_key, $meta_value ) {
+		$object_id = $object->get_id();
 		if ( in_array( $meta_value, array( array(), '' ), true ) && ! in_array( $meta_key, $this->must_exist_meta_keys, true ) ) {
-			$updated = delete_post_meta( $object->get_id(), $meta_key );
+			// Performance note: skip the DELETE query when the meta key is absent — metadata_exists() is cache-aware and avoids a SQL round-trip.
+			$updated = metadata_exists( 'post', $object_id, $meta_key ) && delete_post_meta( $object_id, $meta_key );
```
Comment (kmanijak) [inline plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php]: By avoiding `delete_post_meta` we skip calling `delete_post_metadata` filter ([here](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/meta.php#L448C1-L449C1)) and while it's probably safe for products, this is also extended by coupons and may be extended by other post types.

We should probably check whether there are `delete_post_metadata` callbacks and only skip deletion if there's none.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php [hunk 3]
```diff
@@ -1137,11 +1137,17 @@ protected function update_downloads( &$product, $force = false ) {
...
 	 */
 	protected function update_version_and_type( &$product ) {
-		$old_type = WC_Product_Factory::get_product_type( $product->get_id() );
-		$new_type = $product->get_type();
+		$product_id = $product->get_id();
+		$old_type   = WC_Product_Factory::get_product_type( $product_id );
+		$new_type   = $product->get_type();
+
+		// Skip wp_set_object_terms() when the type is unchanged — it always clears the term cache even on no-op writes.
+		$stored_type_terms = get_the_terms( $product_id, 'product_type' );
```
Comment (kmanijak) [inline plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: Given we `get_the_terms` anyway, let's use `stored_type_terms` to get the "old" product type as apparently `WC_Product_Factory::get_product_type( $product_id );` can return the cached (stale) type.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php [hunk 4]
```diff
@@ -1137,11 +1137,17 @@ protected function update_downloads( &$product, $force = false ) {
...
+		if ( $old_type !== $new_type || empty( $stored_type_terms ) || ! is_array( $stored_type_terms ) ) {
+			wp_set_object_terms( $product_id, $new_type, 'product_type' );
```
Comment (kmanijak) [inline plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:1179]: This avoids calling `set_object_terms` action ([here](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/taxonomy.php#L2995-L3007)). Similar to previous comment I think we should verify if there's callback registered first.

Applicable to variation product type too.

**Review by kmanijak — CHANGES_REQUESTED** [review]
Thanks for this PR. Benchmark stats look promising! But at the same time I'd be super cautious at touching so crucial and well-established paths that external code may heavily rely on.

I left couple of comments that I think would be breaking changes but there may be more.

File: plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php (hunk 2 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php]: Good point; let me explore how to proceed here: I can potentially override this method for product data stores and call it a day if extensions are confirmed not to be affected. This method can be reverted then.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 4 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:1179]: I'll verify, but to clarify: we are not avoiding the callback invocation - only the invocation frequency changes. So technically, the contract is still there.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 3 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: Interesting, I'll dig into this.

File: plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php (hunk 2 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php]: Addressing in 8cdb9dd810e0206b90dbe9a420cab870f4481c7b, efb8f6be9587f74f2eb16d032bfc399fc13c4e69 and 7c751fdc9f6b3e07e9a4fdb4d3c00f0f1c1e9fe9 :

- moved the optimization into override on product data store level (override inherited by all core product data stores)
- verified possible regressions via `delete_post_metadata` (product context) on `all-plugins`: none

The regression risk is low (non-zero, but `all-plugins` analysis reveals none confirmed), and now limited to products only. We can post dev. advisory if we need to boost our confidence (given variable data store audit PRs, we can post a nice advisory actually).

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 3 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: Ok, I like this one - we'll remove a lot of layers/complexity here; it's basically extra free optimization.
I'll verify the number of SQL queries with this change tomorrow and update the PR if it doesn't spawn any extra SQL queries.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php [hunk 5]
```diff
@@ -695,6 +695,35 @@ protected function read_downloads( &$product ) {
...
+	 * @param mixed   $meta_value Value to save.
+	 *
+	 * @return bool
+	 */
+	protected function update_or_delete_post_meta( $product, $meta_key, $meta_value ) {
+		// Performance note: The \WC_Data_Store_WP::update_or_delete_post_meta method is nearly identical, ensure to keep the in sysnc.
+		$product_id = $product->get_id();
+		if ( in_array( $meta_value, array( array(), '' ), true ) && ! in_array( $meta_key, $this->must_exist_meta_keys, true ) ) {
+			$updated = metadata_exists( 'post', $product_id, $meta_key ) && delete_post_meta( $product_id, $meta_key );
+		} else {
```
[the remark this answers is not in the document]
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:720]: Yes, read through this thread/conversation, please: https://github.com/woocommerce/woocommerce/pull/66966#discussion_r3657276971

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 4 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:1179]: Ok, I did a sweep to challenge my own statement (looked into the code and how extensions from `all-plugins` interact with hooks from this function) - it still holds. Contract remains fulfilled - these will be fired on the terms change.

Though I can add a doc-block `@since 11.1.0 one-liner describing the frequency change due to gating updates`

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:720]: Ok, I'll revisit this thread tomorrow with a fresh head.

File: plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php (hunk 2 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php]:
```php
$updated = ( metadata_exists( 'post', $product_id, $meta_key ) || has_filter( 'delete_post_metadata' ) ) && delete_post_meta( $product_id, $meta_key );
```

`delete_post_metadata` is hooked by some extension to track its own meta deletion or to invalidate its own caches. But one of them is subs, a popular one - meaning accounting for the filter will render the optimization low-impact on scale.

Blending all together, it doesn't make sense to add `has_filter( 'delete_post_metadata' )` and we are better off with `metadata_exists()` guard only.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:720]: See https://github.com/woocommerce/woocommerce/pull/66966#discussion_r3659526823

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 3 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: Addressed in 4803106fab316efdfa160526feb42d2bc2ebc559

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 4 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:1179]: Added the tags with clarifications in d09f653dadb8ac39e4ff8e70064b308920f215a7

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 3 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: One case I'm unsure about is variable -> any other type transition: variable product has the type cleared in `\WC_Product_Variation_Data_Store_CPT::update_version_and_type`. With the latest version, the following will not fire:

```php
		if ( $old_type !== $new_type ) {
			$this->updated_props[] = 'product_type';
			do_action( 'woocommerce_product_type_changed', $product, $old_type, $new_type );
		}
```

So we'll need to replicate more bits from `WC_Product_Factory::get_product_type( $product_id )`, to the point it's easier to use `WC_Product_Factory::get_product_type( $product_id )`. WDYT?

File: plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php (hunk 2 above)
Comment (kmanijak) [reply plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php]: I see what you mean. But let's cover that we're decreasing the frequency of metadata deletion in dev advisory 👍

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 3 above)
Comment (kmanijak) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: > variable -> any other type transition: variable product has the type cleared in `\WC_Product_Variation_Data_Store_CPT::update_version_and_type`

If I get it right, this is only related to _variations_ of variable product not _variable products_ themselves. So I don't think this is valid concern 🤔 or I'm missing the point 😅

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 4 above)
Comment (kmanijak) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:1179]: I think we can treat is similarly to previous comment. I get the explanation but I'd cover the information about decreased frequency in dev advisory.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 3 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: > If I get it right, this is only related to variations of variable products, not variable products themselves.

Correct

> So I don't think this is valid concern 🤔 or I'm missing the point 😅

Gotcha.  I have a gap in the possible product-type mutation scenarios (which AI is not able to fill) and am double-checking to better understand those transitions. That's it.

Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: @kmanijak: Apart from this clarification and dev. advisory pending, does anything else in this PR need further work?

**Review by kmanijak — APPROVED** [review]
I did additional smoke testing and it all looks good!

I'm approving the PR given:
- we accepted reduced frequency of hooks,
- there will be a dev advisory documenting it,
- and as we agreed elsewhere, let's push it towards 11.2 to give more time for smoke testing internally.

Great work and great gains! 💪

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 3 above)
Comment (kalessil) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php]: Sadly, we have to switch back to `\WC_Product_Factory::get_product_type(...)` @kmanijak :
- AIRR caught a regression (same convo as I had concerns about) - https://github.com/woocommerce/woocommerce/issues/67396
- fixed in 8ebee975c112307e992a9938f19e668994d71ad7 (with edge-case test)

**Comment by kalessil** [comment]
@coderabbitai: fresh review please

File: plugins/woocommerce/src/Internal/Utilities/ProductUtil.php [hunk 6]
```diff
@@ -54,6 +54,7 @@ public function delete_product_transients_for_products( array $product_ids ): vo
 
 		// Kept for compatibility, WooCommerce core doesn't use product transient versions anymore.
 		\WC_Cache_Helper::get_transient_version( 'product', true );
+		\WC_Cache_Helper::get_transient_version( 'product_query', true );
```
Comment (kalessil) [inline plugins/woocommerce/src/Internal/Utilities/ProductUtil.php:57]: WOOAIRR-40: discovered and fixed an accidental coupling case when unconditional `set_object_terms` triggered this transient refresh. Placing it where it belongs.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php [hunk 7]
```diff
@@ -1132,16 +1161,25 @@ protected function update_downloads( &$product, $force = false ) {
...
-		$old_type = WC_Product_Factory::get_product_type( $product->get_id() );
-		$new_type = $product->get_type();
+		$product_id        = $product->get_id();
+		$old_type          = \WC_Product_Factory::get_product_type( $product_id );
+		$new_type          = $product->get_type();
+		$stored_type_terms = get_the_terms( $product_id, 'product_type' );
+		$stored_type_slug  = ! empty( $stored_type_terms ) && is_array( $stored_type_terms ) ? $stored_type_terms[0]->slug : null;
+
+		// Skip wp_set_object_terms() when the stored term already matches — it always clears the term cache even on no-op writes.
+		if ( $stored_type_slug !== $new_type ) {
```
Comment (kalessil) [inline plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:1178]: Addresses original staleness`\WC_Product_Factory::get_product_type()` concerns.
`\WC_Product_Factory::get_product_type()` was resurercted to address #67396

**Review by kmanijak — APPROVED** [review]
Thanks for additional fixes we missed earlier. This should be good after taking another round of look 🙌
